### ISSUE TITLE
Hyperparameter settings on IEMOCAP dataset

### ISSUE THREAD

**Issue by showgood163** [yaohungt/Multimodal-Transformer#7]
Hi all!

Thanks for sharing the source code! However, I encountered some problems when reproducing the result on IEMOCAP dataset. Any kind of help is appreciated!

I find some suggestions on https://github.com/yaohungt/Multimodal-Transformer/issues/5. Based on that, I fixed all the parameters mentioned in the paper and searched for {embed_dropout,res_dropout,layers}. The model with best validation loss can get ~85% *unweighted* average acc/f1 on validation set but ~81% *unweighted* average acc/f1 on test set, which I think is not close to the reported result in the paper(~84.5%).

Is there something I'm missing or wrongly understand?

Edit: fixed the wrong word (weighted -> unweighted).

**Comment by yaohungt** [comment]
Hi, 

I rechecked the implementation, and I can reproduce the results with our default parameters. 

However, please note that the label distribution for IEMOCAP is skewed, and thus predicting all zeros can already give you a high prediction score. The problem is the same for other datasets with emotion prediction. 

I would recommend only report the numbers for MOSEI sentiment (not recommend MOSI). MOSEI/MOSI are both tiny datasets, but at least MOSEI is less vulnerable to overfitting. Nevertheless, they are already the best datasets I can find for human multimodal language. And this is the only reason I put them in the paper. 

When reproducing the results on IEMOCAP, you can also try more hyperparameters on the number of attention layers. I believe I make it very easy to tune in our model. 

I'm closing the issue, but please do not hesitate to raise any more question you have.

**Comment by showgood163** [comment]
Hi,

Sorry for the wrong word for the metric I choose, I was about to say *unweighed* average acc/f1 not *weighted*, and thanks for explaining that.

I have 2 problems for now:

1. Why the processed data has the sequence length of 20 for IEMOCAP, which is different from both MOSI and MOSEI (which is 50)?  

2. When you say *default* parameters, which are you talking about? The parameters provided by the code or the paper? Could you explain that in detail?

And, if you dont mind, [Here](https://github.com/showgood163/Multimodal-Transformer/blob/master/main.py) are the parameter settings I'm using for now. I also changed the [Transformers Hidden Unit Size](https://github.com/yaohungt/Multimodal-Transformer/blob/master/src/models.py#L15) to 40 but this is not listed in the main.py so I think I may do something wrong here.

**Comment by yaohungt** [comment]
1. You may want to check the Multimodal SDK, which is not designed by me. You can also get in touch with Paul, who did the data preprocessing for our work.

2. I'm using the parameters mentioned in the paper. But the performance varies a lot from different random seeds.

Changing the hidden unit size may have an effect on the final performance. However, we find changing dropout and number of layers affects more.

**Comment by showgood163** [comment]
Hi,

I think `#epochs to decay learning rate` is an important hyper parameter, and that is not clear in the paper. Could you tell me what you set for IEMOCAP, MOSI and MOSEI dataset?

**Comment by yaohungt** [comment]
Can you point out which part is you referring to?

**Comment by showgood163** [comment]
Hi,

The `#epochs to decay learning rate` is in https://github.com/yaohungt/Multimodal-Transformer/blob/master/main.py#L64.

Also, could you describe in detail how you perform grid search, like what hyper parameters do you fix and what hyper parameters do you search? That will be really helpful!

Thanks!

**Comment by yaohungt** [comment]
We do not change this hyper-parameter. And the rest of the hyper-parameters is provided in Supplementary. 

Moreover, we do not actually perform a grid search. The hyper-parameters are set with only a few trials. 

We strongly believe that you can perform your own favorite grid search, and the performance should be higher than reported.

**Comment by showgood163** [comment]
Sorry but I have to ask a question (like a fool), I'm running this code with the latest pytorch (1.5.0/1.4.0), will that be a problem? And what should I do to get a result that is close to which reported in the paper?

**Comment by yaohungt** [comment]
I don't think it would be a problem with different PyTorch versions. 

You may test with different combinations of hyper-parameters to get better performance.

**Comment by showgood163** [comment]
Hi,

I have questions about the results you reported, i.e. MOSI/MOSEI, are they come from a single model or many models? I never get a model with every metric being good, espcially Acc7 and Acc2.

Also, how do you deal with the overfitting on the evaluation set? If the model gets a good generalization in a certain epoch but after some time it overfits on the evaluation set and get worse performance on the test set, then what will you report?

**Comment by yaohungt** [comment]
1. Single model.
2. You should report the best number on the evaluation set. You may also use early stopping to prevent overfitting.

**Comment by showgood163** [comment]
So when the model reached new best valid loss, I should report the corresponding results on the eval set not on the test set right?

**Comment by yaohungt** [comment]
This is the standard way of reporting the results.

**Comment by showgood163** [comment]
But why not report the results on the test set? The model is supposed to generalize well.

**Comment by yaohungt** [comment]
You report the number on the test set according to the trial that reaches the best performance on the validation set.

**Comment by showgood163** [comment]
Another question: did you use the weight decay when training this model?

**Comment by yaohungt** [comment]
No, we don't. And actually, I encourage you to read the code since more details are in the code.
